src/GovUk.Education.ExploreEducationStatistics.Admin/Services/Public.Data/DataSetService.cs

```diff
@@ -170,66 +184,56 @@ private async Task<DataSetViewModel> MapDataSet(DataSet dataSet, CancellationTok
-            contentDbContext,
-            previousReleaseFileIds,
-            cancellationToken);
-    }
-
-    private static async Task<List<Guid>> GetReleaseIdsForReleaseFiles(
-        ContentDbContext contentDbContext,
-        List<Guid> releaseFileIds,
+    private async Task<Dictionary<Guid, Dictionary<Guid, ReleaseVersion>>> GetPreviousReleasesByVersionByDataSet(
+        IReadOnlyList<DataSet> dataSets, 
         CancellationToken cancellationToken)
     {
-        return await contentDbContext
+        var dataSetsByPreviousReleaseFileId = dataSets
+            .SelectMany(ds => ds.Versions)
+            .ToDictionary(dsv => dsv.ReleaseFileId, dsv => dsv.DataSet);
+
+        var releasesByReleaseFileId = await contentDbContext
             .ReleaseFiles
-            .Where(releaseFile => releaseFileIds.Contains(releaseFile.Id))
-            .Select(releaseFile => releaseFile.ReleaseVersion.ReleaseId)
-            .ToListAsync(cancellationToken);
+            .Include(rf => rf.ReleaseVersion)
+            .Where(releaseFile => dataSetsByPreviousReleaseFileId.Keys.Contains(releaseFile.Id))
+            .ToDictionaryAsync(
+                rf => rf.Id,
+                rf => rf.ReleaseVersion,
+                cancellationToken);
+
+        return releasesByReleaseFileId
+            .Select(d => new
+            {
+                ReleaseFileId = d.Key,
+                ReleaseVersion = d.Value,
+                DataSet = dataSetsByPreviousReleaseFileId[d.Key]
+            })
+            .GroupBy(a => a.DataSet)
+            .ToDictionary(
+                g => g.Key.Id,
+                g => g.ToDictionary(
+                        a => a.DataSet.Versions.Single(v => v.ReleaseFileId == a.ReleaseFileId).Id,
+                        a => a.ReleaseVersion
+                    )
+            );
     }
 
-    private async Task<IReadOnlyDictionary<DataSetVersion, ReleaseFile>> GetReleaseFilesByDataSetVersion(
-        DataSet dataSet,
+    private async Task<IReadOnlyDictionary<Guid, ReleaseFile>> GetDataSetPreviousReleaseFilesByVersion(
+        DataSet dataSet, 
         CancellationToken cancellationToken)
     {
-        if (dataSet.LatestDraftVersion is null && dataSet.LatestLiveVersion is null)
-        {
-            return new Dictionary<DataSetVersion, ReleaseFile>();
-        }
-
-        var dataSetVersions = new List<DataSetVersion>();
-        var predicate = PredicateBuilder.False<ReleaseFile>();
-
-        if (dataSet.LatestDraftVersion is not null)
-        {
-            dataSetVersions.Add(dataSet.LatestDraftVersion);
-
-            predicate = predicate.Or(rf => rf.Id == dataSet.LatestDraftVersion.ReleaseFileId);
-        }
-
-        if (dataSet.LatestLiveVersion is not null)
-        {
-            dataSetVersions.Add(dataSet.LatestLiveVersion);
-
-            predicate = predicate.Or(rf => rf.Id == dataSet.LatestLiveVersion.ReleaseFileId);
-        }
+        var dataSetVersionsByPreviousReleaseFileId = dataSet
```

Comment:
I'm not sure what makes these `previous`.  Can it not also contain the latest version's ReleaseFile as well?  In which case, maybe it needs a slight rename to `GetReleaseFilesByDataSetVersionId`

Reply:
Now that you mention it, this doesn't make sense to me either!

I've renamed things to make it a bit clearer :)

Reply:
Cool!